### PR TITLE
VZ-6460: Automatically pull CA cert secret from managed cluster

### PR DESCRIPTION
**Pull request by fredtibbitts**
This PR adds processing to the VMC reconcile logic that:

1. Waits for the managed cluster to become active in Rancher
2. Reaches into the managed cluster (using the Rancher API proxy) and gets the CA cert secret
3. Creates/updates the CA cert secret on the admin cluster
4. Sets the `caSecret` field in the VMC spec to reference the secret

I tested this locally in a multi-cluster setup by:

1. Created an empty VMC on the admin cluster
2. Navigated to the Rancher UI and saw that the managed cluster was pending
3. Copied the command to register the cluster
4. Ran the command on the managed cluster
5. Examined the VPO logs and saw that it pulled the CA secret
6. Described the VMC and saw that the `caSecret` field was now populated
7. Dumped the contents of the CA secret

Included in this work was some refactoring in `rancher.go` to be able to re-use the code to login to Rancher and create a session token so that it could be used to make multiple calls.

### CODE REVIEW

File: platform-operator/controllers/clusters/rancher.go [hunk 1]
```diff
@@ -36,13 +37,18 @@ const (
...
 	clusterRegTokenPath = "/v3/clusterregistrationtoken" //nolint:gosec //#gosec G101
 	manifestPath        = "/v3/import/"
 	loginPath           = "/v3-public/localProviders/local?action=login"
 
+	k8sClustersPath = "/k8s/clusters/"
+
 	// this host resolves to the cluster IP
 	nginxIngressHostName = "ingress-controller-ingress-nginx-controller.ingress-nginx"
+
+	clusterStateActive = "active"
```
Comment (desagar) [inline platform-operator/controllers/clusters/rancher.go:51]: minor nit - should we call this rancherClusterStateActive?

File: platform-operator/controllers/clusters/rancher.go [hunk 2]
```diff
@@ -233,7 +243,99 @@ func getClusterIDFromRancher(rc *rancherConfig, clusterName string, log vzlog.Ve
 	}
 
 	return httputil.ExtractFieldFromResponseBodyOrReturnError(responseBody, "data.0.id", "unable to find clusterId in Rancher response")
+}
+
+// isManagedClusterActive returns true if the managed cluster is active
+func isManagedClusterActive(rc *rancherConfig, clusterID string, log vzlog.VerrazzanoLogger) (bool, error) {
```
Comment (desagar) [inline platform-operator/controllers/clusters/rancher.go:249]: similar to above minor nit - should we call this isManagedClusterActiveInRancher?

File: platform-operator/controllers/clusters/rancher.go [hunk 3]
```diff
@@ -233,7 +243,99 @@ func getClusterIDFromRancher(rc *rancherConfig, clusterName string, log vzlog.Ve
...
+	// Rancher temporarily sets the state of a new cluster to "active" before setting it to "pending", so we also check for the "agentImage" field
+	// to know that the cluster is really active
+	return state == clusterStateActive && len(agentImage) > 0, nil
+}
+
+// getCACertFromManagedCluster attempts to get the CA cert from the managed cluster using the Rancher API proxy. It first checks for
+// the Rancher TLS secret and if that is not found it looks for the Verrazzano system TLS secret.
+func getCACertFromManagedCluster(rc *rancherConfig, clusterID string, log vzlog.VerrazzanoLogger) (string, error) {
+	// first look for the Rancher TLS secret
+	caCert, err := getCACertFromManagedClusterSecret(rc, clusterID, rancherNamespace, cons.AdditionalTLS, "ca-additional.pem", log)
```
Comment (desagar) [inline platform-operator/controllers/clusters/rancher.go:281]: there's a constant for the "ca-additional.pem" as well - AdditionalTLSCAKey

File: platform-operator/controllers/clusters/rancher.go [hunk 4]
```diff
@@ -233,7 +243,99 @@ func getClusterIDFromRancher(rc *rancherConfig, clusterName string, log vzlog.Ve
...
+	if err != nil {
+		return "", err
+	}
+
+	if caCert != "" {
+		return caCert, nil
+	}
+
+	// didn't find the Rancher secret so next look for the verrazzano-tls secret
+	caCert, err = getCACertFromManagedClusterSecret(rc, clusterID, cons.VerrazzanoSystemNamespace, constants.VerrazzanoIngressSecret, "ca.crt", log)
```
Comment (desagar) [inline platform-operator/controllers/clusters/rancher.go:291]: there is a constant - mcconstants.CaCrtKey = "ca.crt"

**Review by desagar — APPROVED** [review]

File: platform-operator/controllers/clusters/rancher.go (hunk 1 above)
Comment (fredtibbitts) [reply platform-operator/controllers/clusters/rancher.go]: Renamed.

File: platform-operator/controllers/clusters/rancher.go (hunk 2 above)
Comment (fredtibbitts) [reply platform-operator/controllers/clusters/rancher.go]: Renamed.

File: platform-operator/controllers/clusters/rancher.go (hunk 3 above)
Comment (fredtibbitts) [reply platform-operator/controllers/clusters/rancher.go]: Done.

File: platform-operator/controllers/clusters/rancher.go (hunk 4 above)
Comment (fredtibbitts) [reply platform-operator/controllers/clusters/rancher.go]: Done.

**Review by desagar — APPROVED** [review]

**Pull request merged**